interfaces/builtin/service_watchdog.go

```diff
@@ -0,0 +1,78 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+)
+
+const serviceWatchdogSummary = `allows use of systemd service watchdog`
+
+const serviceWatchdogBaseDeclarationSlots = `
+  service-watchdog:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const serviceWatchdogConnectedPlugAppArmorTemplate = `
+# Allow sending notification messages to systemd through the notify socket
+"{{notify-socket}}" w,
+`
+
+type serviceWatchdogInterface struct {
+	commonInterface
+}
+
+var osGetenv = os.Getenv
+
+func (iface *serviceWatchdogInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	notifySocket := osGetenv("NOTIFY_SOCKET")
+	if notifySocket == "" {
+		notifySocket = "/run/systemd/notify"
+	} else {
+		// must be an absolute path without any AppArmor regular
+		// expression (AARE) characters or double quotes
+		if !strings.HasPrefix(notifySocket, "/") || strings.ContainsAny(notifySocket, `?*[]{}^"`) {
```

Comment:
This looks good from a policy POV. It would be nice if we could share this check with overlay.go since it is an exact code copy with just the checked string as different, but that can be done in a followup PR. Could do (maybe in interfaces/apparmor/apparmor.go?):
```
func IsValidAppArmorFilePath(path string) error {
  	// path must be an absolute path without any AppArmor regular expression
	// (AARE) characters or double quotes to be considered
	if !strings.HasPrefix(path, "/") || strings.ContainsAny(path, `?*[]{}^"`) {
		return fmt.Errorf("cannot use %q as AppArmor file path", path)
	}
	return nil
  }
```

Not a blocker for this PR, but I wouldn't mind it. :)

Reply:
Briefly discussed during standup. I'll probably open a follow-up PR to move this to another package.